src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -105,62 +108,180 @@ public void run() {
+            this.nsdManager,
+            serviceInfo,
+            callbackHandle,
+            contextHandle,
+            chipMdnsCallback,
+            timeoutRunnable,
+            multicastLock,
+            mainThreadHandler,
+            nsdManagerResolverAvailState);
+    serviceFinderResolver.start();
+
+    Executors.newSingleThreadScheduledExecutor()
+        .schedule(
+            new Runnable() {
+              @Override
+              public void run() {
+                nsdManager.stopServiceDiscovery(serviceFinderResolver);
+                multicastLock.release();
+              }
+            },
+            BROWSE_SERVICE_TIMEOUT,
+            TimeUnit.MILLISECONDS);
+
+    mainThreadHandler.postDelayed(timeoutRunnable, RESOLVE_SERVICE_TIMEOUT);
+  }
+
+  private static class NsdServiceFinderAndResolver implements NsdManager.DiscoveryListener {
+    private static final String TAG = NsdServiceFinderAndResolver.class.getSimpleName();
+
+    private final NsdManager nsdManager;
+    private final NsdServiceInfo targetServiceInfo;
+    private final long callbackHandle;
+    private final long contextHandle;
+    private final ChipMdnsCallback chipMdnsCallback;
+    private final Runnable timeoutRunnable;
+    private final MulticastLock multicastLock;
+    private final Handler mainThreadHandler;
+    @Nullable private final NsdManagerResolverAvailState nsdManagerResolverAvailState;
+
+    public NsdServiceFinderAndResolver(
+        final NsdManager nsdManager,
+        final NsdServiceInfo targetServiceInfo,
+        final long callbackHandle,
+        final long contextHandle,
+        final ChipMdnsCallback chipMdnsCallback,
+        final Runnable timeoutRunnable,
+        final MulticastLock multicastLock,
+        final Handler mainThreadHandler,
+        final NsdManagerResolverAvailState nsdManagerResolverAvailState) {
+      this.nsdManager = nsdManager;
+      this.targetServiceInfo = targetServiceInfo;
+      this.callbackHandle = callbackHandle;
+      this.contextHandle = contextHandle;
+      this.chipMdnsCallback = chipMdnsCallback;
+      this.timeoutRunnable = timeoutRunnable;
+      this.multicastLock = multicastLock;
+      this.mainThreadHandler = mainThreadHandler;
+      this.nsdManagerResolverAvailState = nsdManagerResolverAvailState;
     }
 
-    this.nsdManager.resolveService(
-        serviceInfo,
-        new NsdManager.ResolveListener() {
-          @Override
-          public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
-            Log.w(
-                TAG,
-                "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
-            chipMdnsCallback.handleServiceResolve(
-                instanceName, serviceType, null, null, 0, null, callbackHandle, contextHandle);
+    public void start() {
+      this.nsdManager.discoverServices(
+          targetServiceInfo.getServiceType(), NsdManager.PROTOCOL_DNS_SD, this);
```

Comment:
do we need to acquire the multicast lock here so we don't try to initiate service discovery more than once?

Reply:
It is acquired. I've moved it to be closer to the call to startServiceDiscovery now to make it more obvious.